### PR TITLE
enableWithArtifact and enableWithClasspath

### LINKED ISSUE #57: Enhanced enableIf for arbitrary jar dependencies
``` scala
  @enableIf(c =>
    c.classPath.exists(
      _.getPath.matches(".*spark-catalyst_2\\.\\d+-3\\.2\\..*")
    )
  )
```

@Renkai is using enableIf to conditional compile a Spark-based project here: https://github.com/eto-ai/rikai/pull/531

Using classpath and regex, it works fine. It would be nice to have a special macro like `enableClasspath`.

### PR DESCRIPTION
**Pull request by da-tubi**
Co Authored With @Renkai
closes #57

### CODE REVIEW

File: src/main/scala/com/thoughtworks/enableWithArtifact.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import scala.reflect.internal.annotations.compileTimeOnly
+import scala.reflect.macros.Context
+import scala.util.matching.{Regex, UnanchoredRegex}
+
+
+object enableWithArtifact {
+
+  def checkArtifact(c: Context, artifactId: String, regex: Regex): Boolean = {
+    c.classPath.exists(
+      _.getPath.matches(s".*${artifactId}-${regex.toString}")
```
Comment (da-liii) [inline src/main/scala/com/thoughtworks/enableWithArtifact.scala:13]: regex could be used to identify classifiers, non jar resources

**Comment by Atry** [comment]
Would you mind updating README as well?

**Comment by da-liii** [comment]
OK, let me add a new section and then update the `Motivation` section.

**Comment by Atry** [comment]
This PR seems inconsistent with existing annotations, as the existing annotations `enableMembersIf` applies on all members, while this PR did not provide such ability.

**Comment by da-liii** [comment]
Let me make another PR later.

**Comment by da-liii** [comment]
This pr is ready, see the updated README here: https://github.com/darcy-shen/enableIf.scala/tree/feature/enableWithClasspath

**Comment by da-liii** [comment]
> This PR seems inconsistent with existing annotations, as the existing annotations enableMembersIf applies on all members, while this PR did not provide such ability.

Let me check if there is any possibility of merging `enableMembersIf` into `enableIf`? Wish to receive some help from you if any.

**Comment by Atry** [comment]
> Let me check if there is any possibility of merging `enableMembersIf` into `enableIf`? Wish to receive some help from you if any.

It does not sound like a good idea because you are essentially trying to do different things in a single function, which is bad.


How about this:

``` scala
def hasArtifactInClasspath(artifactId: String, version: String)(context: Context): Boolean
```

Then the user will be able to use it either like this

``` scala
@enableMembersIf(hasArtifactInClasspath("scala-library", scala.util.Properties.versionNumberString))
```
or 
``` scala
@enableIf(hasArtifactInClasspath("scala-library", scala.util.Properties.versionNumberString))
```

**Comment by da-liii** [comment]
Thanks for your guide, let me create another pull request!

**Comment by da-liii** [comment]
I tried in #70 , the syntax does not compile:

``` scala
@enableIf(hasArtifactInClasspath("scala-library", scala.util.Properties.versionNumberString))
```

**Comment by da-liii** [comment]
In my opinion, `@enableIf` is used more frequently than `@enableMembersIf`. For example, in https://github.com/eto-ai/rikai/pull/531 , to make the code compile with both Apache Spark 3.1.2 and 3.2.0, we use the (almost) same pattern of `@enableIf` for 6 times.

That's why I introduce the simplified `@enableWithClasspath` and `@enableWithArtifact` here, just to make it much more readable.

Once we need `@enableMembersWithClasspath` and `@enableMembersWithArtifact`, we could add it later. But I think `@enableMembersIf` is seldom adopted. We do not need the simplified version of `@enableMembersIf`.

File: src/main/scala/com/thoughtworks/enableWithArtifact.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    )
+  }
+
+  def checkArtifact(c: Context, artifactId: String, version: String): Boolean = {
+    val versionRegex = version.replace(".", "\\.")
+    c.classPath.exists(
+      _.getPath.matches(s".*${artifactId}-${versionRegex}.*")
```
Comment (da-liii) [inline src/main/scala/com/thoughtworks/enableWithArtifact.scala:20]: is it a good idea to add one more predict for scala style artifactId

```
_.getPath.matches(s".*${artifactId}-${versionRegex}.*") ||
_.getPath.matches(s".*${artifactId}_${scalaMajorVersion}-${versionRegex}.*")
```

scalaMajorVersion is determined by the compiler Scala version, like 2.12/2.13

Comment (da-liii) [reply src/main/scala/com/thoughtworks/enableWithArtifact.scala:20]: In this way, for scala deps:

```
@enableWithArtifact("spark-catalyst", "3.1.2")
```
We do not need to hard code `"spark-catalyst"` as `"spark-catalyst_2.12"` or `"spark-catalyst_2.13"`

Comment (da-tubi) [reply src/main/scala/com/thoughtworks/enableWithArtifact.scala:20]: A new pr with that improvements is here: https://github.com/ThoughtWorksInc/enableIf.scala/pull/71

**Pull request closed without merging**